code/modules/GOAP/GOAPPlanner.dm

```diff
@@ -0,0 +1,107 @@
+
+/datum/goap_planner
+	var/list/stored_nodes = list()
+
+
+/datum/goap_planner/proc/Plan(atom/agent, list/actions, list/worldstate, list/goal)
+	for(var/a in actions)
+		var/datum/goap_action/GA = a
+		GA.Reset()
+
+	var/list/usable_actions = list()
+
+	for(var/a in actions)
+		var/datum/goap_action/GA = a
+		if(GA.OnCooldown(agent, worldstate))
+			continue
+		if(GA.AdvancedPreconditions(agent, worldstate))
+			usable_actions += GA
+
+	//Oh god, trees! I hate this!
+
+	var/list/plan_tree = list()
+
+	var/datum/goap_plan_node/start = new()
+	start.state = goal
+	var/list/end = worldstate
+
+	plan_tree = BuildPossiblePlans(end, start, usable_actions, goal)
+
+	if(!plan_tree.len)
+		return null
+
+	//Cheapest path
+	var/datum/goap_plan_node/cheapest
+	for(var/node in plan_tree)
+		var/datum/goap_plan_node/N = node
+		if(!cheapest || N.cost < cheapest.cost)
+			cheapest = N
+	//Go up the tree, from the cheapest bottom leaf
+	var/list/climblist = list()
+	var/datum/goap_plan_node/climber = cheapest
+	while(climber)
+		if(climber.action)
+			climblist += climber.action
+		climber = climber.parent
+	//Reverse the plan
+	var/list/plan = list()
+	for(var/i = climblist.len to 1 step -1)
+		plan += climblist[i]
+	return plan
+
+
+//Works backwards from the goal to the start to find the best path
+//TODO: remove recursion
```

Comment:
using recursion here is likely to trigger byond's recursion detection, which is real twitchy

Reply:
Dunno an iterative solution to this proc yet and it works fine for now but I’m sure there’d be performance gains if recursion was removed

Reply:
its not a performance gain thing, its a "byond will sometimes decide theres too much recursion happening and kill the proc"

Reply:
How do i fix that?

Reply:
During testing we never had byond infinite loop proc

Reply:
Can confirm never triggered the recursion detection, and there's definitely some performance to gain.

I've not written DM for ages (and BuildPossiblePlans was partly rewritten from when I wrote it), but I believe this is a decent iterative version.

```dm
// tg already has a pop() in _lists.dm, but not a peek()
#define PEEKLIST(L) ((L.len ? L[L.len] : null))
		
/*
	Returns a list of plan nodes that are the END of a possible plan
	These nodes have the cost of the entire plan (allowing you to choose, eg: the cheapest plan)
	to construct a plan from the chosen node, simply follow the node parent
	chain until you hit a node with no parent:
	
	var/list/plan = list()
	var/datum/goap_plan_node/climber = chosen_node
	while(climber)
		if(climber.action)
			plan += climber.action
		climber = climber.parent
	... then reverse the plan list to get it in START->END order ...
*/
/datum/goap_planner/proc/BuildPossiblePlans(list/end, datum/goap_plan_node/parent0, list/usable_actions0, list/goal)
	var/list/plan_end_nodes  = list()                 //List of nodes that end a plan (thus, can be used to form plans in reverse)
	
	//Stacks
	var/list/parent_stack    = list(parent0)          //Parent node stack (List of nodes)
	var/list/actions_stack   = list(usable_actions0)  //Actions stack (List of lists of actions)
	
	var depth = 1
	while(depth>0)	
		var/list/usable_actions = PEEK(actions_stack)
		for(var/a in usable_actions)
			var/datum/goap_action/GA = a
			var/datum/goap_plan_node/parent = PEEKLIST(parent_stack)
			if(!InState(GA.effects, parent.state, "usable_actions")) // this won't get us to the goal
				continue
				
			//What does the world look like if we run this action?
			var/list/current_state = ShowMeTheFuture(parent.state, GA.effects, GA.preconditions) // remove the effects and add the preconditions
			var/datum/goap_plan_node/node = new()
			node.parent = parent
			var/budget = parent.cost+GA.cost
			node.cost = budget
			node.state = current_state
			node.action = GA

			//This node reaches the goal, add it, but keep going to find more plans
			if(InState(current_state, end, "add_to_plan_tree"))
				plan_end_nodes += node
			else
				usable_actions -= GA //NOTE 2
				parent_stack += node			
				actions_stack[actions_stack.len+1] = usable_actions.Copy() //NOTE 1
				depth++
			CHECK_TICK
			
		pop(parent_stack)
		pop(actions_stack)		
		depth--
	return plan_end_nodes
```

Note 1:
This also fixes a logic problem, `usable_actions` should really be copied before each plan, otherwise we're not exploring perfectly valid plans, which may end up being the cheapest possible plan, meaning at the moment GOAP AIs may be unable to perform the theoretically cheapest plan.
Fixing this will probably make them even SMARTER/powergamey-er, but it should still be fixed.

Note 2:
This remains, even with the logic fix (Note 1), as we've picked this node to exist as a branch in the tree, and it's unlikely its preconditions are valid at a later position in the plan than here, where it's already succeed. You could try removing this line though, just to see if they get any smarter (I believe it'll just be unnecessary extra work, no extra smartness)

Obviously, this isn't tested, but it should be logically sound.